### ISSUE TITLE
[Kubernetes] Failures when creating workspaces rapidly

### ISSUE THREAD

**Issue by mjshashank** [eclipse-che/che#9787]
### Description

This might be related to [this issue](https://github.com/openshiftio/openshift.io/issues/3348) but I'm getting the same error at a rate much greater than the 10% described there.

My use-case requires a fairly large set-up of Che and this issue occurs when I try to call the API rapidly creating many workspaces.

The following are some of the server logs:

```

2018-05-23 17:59:10,145[io-8080-exec-20]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_ZKCAAJQZ' with id 'workspacewdfqukz7k1xd2sia' created by user 'admin'
2018-05-23 17:59:11,230[io-8080-exec-32]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_DOKRELNG' with id 'workspace4hv17airb3qimhay' created by user 'admin'
2018-05-23 17:59:13,877[io-8080-exec-15]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_IXQUGLXO' with id 'workspaceio7jo9keqbfxkgfk' created by user 'admin'
2018-05-23 17:59:18,562[io-8080-exec-18]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_ZKCAAJQZ' with id 'workspacewdfqukz7k1xd2sia' by user 'admin'
2018-05-23 17:59:19,996[io-8080-exec-16]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_DOKRELNG' with id 'workspace4hv17airb3qimhay' by user 'admin'
2018-05-23 17:59:23,090[io-8080-exec-29]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_IXQUGLXO' with id 'workspaceio7jo9keqbfxkgfk' by user 'admin'
2018-05-23 18:00:28,169[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacey0lsj91u0v8ywply` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:00:28.16756106Z`
2018-05-23 18:00:28,178[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacey0lsj91u0v8ywply` text=`  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current` time=`2018-05-23T18:00:28.178273741Z`
2018-05-23 18:00:28,179[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacey0lsj91u0v8ywply` text=`                                 Dload  Upload   Total   Spent    Left  Speed` time=`2018-05-23T18:00:28.179401594Z`
2018-05-23 18:00:28,767[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b24100 36.6M  100 36.6M    0     0  62.2M      0 --:--:-- --:--:-- --:--:-- 62.3M` time=`2018-05-23T18:00:28.766498614Z`
2018-05-23 18:00:38,276[aceSharedPool-3]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 356]   - Workspace 'admin:h8k1_CBQMBJTM' with id 'workspacey0lsj91u0v8ywply' started by user 'admin'
2018-05-23 18:00:52,565[io-8080-exec-16]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_SGYIFDWX' with id 'workspacevs693305bcc6fx9m' created by user 'admin'
2018-05-23 18:01:01,220[io-8080-exec-18]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_SGYIFDWX' with id 'workspacevs693305bcc6fx9m' by user 'admin'
2018-05-23 18:03:58,596[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacetv18ehhio919fr2c. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacetv18ehhio919fr2c' reached timeout
2018-05-23 18:03:58,768[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_JEETZBBX' with id 'workspacetv18ehhio919fr2c' start failed
2018-05-23 18:04:01,593[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacev1yakqvxs6xsou8a` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:04:01.592450158Z`
2018-05-23 18:04:12,511[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_HURJCNUC' with id 'workspacebdx5d6pl0pgq0swe' created by user 'admin'
2018-05-23 18:04:22,185[io-8080-exec-32]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_HURJCNUC' with id 'workspacebdx5d6pl0pgq0swe' by user 'admin'
2018-05-23 18:05:01,327[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacedpq1xfez1pld5q8k` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:05:01.326426446Z`
2018-05-23 18:05:02,106[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacedpq1xfez1pld5q8k` text=`                                 Dload  Upload   Total   Spent    Left  Speed` time=`2018-05-23T18:05:02.106189181Z`
2018-05-23 18:05:02,106[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacedpq1xfez1pld5q8k` text=`  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current` time=`2018-05-23T18:05:02.104728187Z`
2018-05-23 18:05:02,610[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b24100 36.6M  100 36.6M    0     0  73.4M      0 --:--:-- --:--:-- --:--:-- 73.5M` time=`2018-05-23T18:05:02.609517879Z`
2018-05-23 18:07:58,604[aceSharedPool-2]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacev1yakqvxs6xsou8a. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacev1yakqvxs6xsou8a' reached timeout
2018-05-23 18:07:58,609[aceSharedPool-1]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacemmff8hkhae0ccilo. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacemmff8hkhae0ccilo' reached timeout
2018-05-23 18:07:58,815[aceSharedPool-2]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_KORDVGBQ' with id 'workspacev1yakqvxs6xsou8a' start failed
2018-05-23 18:07:58,892[aceSharedPool-1]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_OQYSHYCB' with id 'workspacemmff8hkhae0ccilo' start failed
2018-05-23 18:08:09,272[io-8080-exec-37]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_KJFCPWTW' with id 'workspace56avnflf4tti1vna' created by user 'admin'
2018-05-23 18:08:12,845[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_XLTDMOMY' with id 'workspacehbrx3w6opa7xzbde' created by user 'admin'
2018-05-23 18:08:18,126[io-8080-exec-20]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_KJFCPWTW' with id 'workspace56avnflf4tti1vna' by user 'admin'
2018-05-23 18:08:21,653[io-8080-exec-36]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_XLTDMOMY' with id 'workspacehbrx3w6opa7xzbde' by user 'admin'
2018-05-23 18:08:58,628[aceSharedPool-3]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacedpq1xfez1pld5q8k. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacedpq1xfez1pld5q8k' reached timeout
2018-05-23 18:08:58,798[aceSharedPool-3]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_FWEQHHZG' with id 'workspacedpq1xfez1pld5q8k' start failed
2018-05-23 18:09:11,910[io-8080-exec-37]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_SWNKKFII' with id 'workspacegtvuzok8iyrco6kw' created by user 'admin'
2018-05-23 18:09:20,664[io-8080-exec-35]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_SWNKKFII' with id 'workspacegtvuzok8iyrco6kw' by user 'admin'
2018-05-23 18:12:58,576[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspaceco3xbquu7gunwtbd. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspaceco3xbquu7gunwtbd' reached timeout
2018-05-23 18:12:58,774[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_TFRQAPIS' with id 'workspaceco3xbquu7gunwtbd' start failed
2018-05-23 18:13:08,368[io-8080-exec-18]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_HYTXPZHV' with id 'workspacej9cvcshgl43vwrsq' created by user 'admin'
2018-05-23 18:13:17,806[io-8080-exec-16]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_HYTXPZHV' with id 'workspacej9cvcshgl43vwrsq' by user 'admin'
2018-05-23 18:16:58,586[aceSharedPool-1]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacewdfqukz7k1xd2sia. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacewdfqukz7k1xd2sia' reached timeout
2018-05-23 18:16:58,608[aceSharedPool-2]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspaced76r1vxwhubfnc3y. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspaced76r1vxwhubfnc3y' reached timeout
2018-05-23 18:16:58,786[aceSharedPool-1]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_ZKCAAJQZ' with id 'workspacewdfqukz7k1xd2sia' start failed
2018-05-23 18:16:58,873[aceSharedPool-2]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_FARJYKPR' with id 'workspaced76r1vxwhubfnc3y' start failed
2018-05-23 18:17:08,629[io-8080-exec-37]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_BLBXHRPG' with id 'workspacedw82mqesqavkse90' created by user 'admin'
2018-05-23 18:17:09,556[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_SFCXLEID' with id 'workspacex78m7wn9gqj2qxmc' created by user 'admin'
2018-05-23 18:17:18,084[io-8080-exec-24]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_BLBXHRPG' with id 'workspacedw82mqesqavkse90' by user 'admin'
2018-05-23 18:17:18,509[io-8080-exec-36]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_SFCXLEID' with id 'workspacex78m7wn9gqj2qxmc' by user 'admin'
2018-05-23 18:17:58,640[aceSharedPool-3]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspace4hv17airb3qimhay. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspace4hv17airb3qimhay' reached timeout
2018-05-23 18:17:58,808[aceSharedPool-3]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_DOKRELNG' with id 'workspace4hv17airb3qimhay' start failed
2018-05-23 18:18:10,656[io-8080-exec-36]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_EMWOKNSU' with id 'workspacencc5o0voqjb5e0mw' created by user 'admin'
2018-05-23 18:18:19,590[io-8080-exec-24]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_EMWOKNSU' with id 'workspacencc5o0voqjb5e0mw' by user 'admin'
2018-05-23 18:21:58,569[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspaceio7jo9keqbfxkgfk. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspaceio7jo9keqbfxkgfk' reached timeout
2018-05-23 18:21:58,775[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_IXQUGLXO' with id 'workspaceio7jo9keqbfxkgfk' start failed
2018-05-23 18:21:59,980[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacevs693305bcc6fx9m` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:21:59.979453837Z`
2018-05-23 18:22:13,421[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_GZUWJOJH' with id 'workspace1vj2gd0qhkfhdywd' created by user 'admin'
2018-05-23 18:22:23,084[io-8080-exec-15]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_GZUWJOJH' with id 'workspace1vj2gd0qhkfhdywd' by user 'admin'
2018-05-23 18:23:12,472[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspace56avnflf4tti1vna` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:23:12.471711688Z`
2018-05-23 18:23:13,042[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspace56avnflf4tti1vna` text=`  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current` time=`2018-05-23T18:23:13.04149316Z`
2018-05-23 18:23:13,043[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `com.domain.che.fsagent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspace56avnflf4tti1vna` text=`                                 Dload  Upload   Total   Spent    Left  Speed` time=`2018-05-23T18:23:13.042549413Z`
2018-05-23 18:25:58,604[aceSharedPool-2]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacebdx5d6pl0pgq0swe. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacebdx5d6pl0pgq0swe' reached timeout
2018-05-23 18:25:58,613[aceSharedPool-1]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacevs693305bcc6fx9m. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacevs693305bcc6fx9m' reached timeout
2018-05-23 18:25:58,781[aceSharedPool-2]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_HURJCNUC' with id 'workspacebdx5d6pl0pgq0swe' start failed
2018-05-23 18:25:58,872[aceSharedPool-1]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_SGYIFDWX' with id 'workspacevs693305bcc6fx9m' start failed
2018-05-23 18:26:08,788[io-8080-exec-36]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_OGMHDHKV' with id 'workspaceypw752pfmuru0y7p' created by user 'admin'
2018-05-23 18:26:09,992[io-8080-exec-32]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_QGJEGLDP' with id 'workspace7nt9hak82wxyblh7' created by user 'admin'
2018-05-23 18:26:18,428[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_OGMHDHKV' with id 'workspaceypw752pfmuru0y7p' by user 'admin'
2018-05-23 18:26:19,234[io-8080-exec-16]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_QGJEGLDP' with id 'workspace7nt9hak82wxyblh7' by user 'admin'
2018-05-23 18:26:58,579[aceSharedPool-3]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspace56avnflf4tti1vna. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspace56avnflf4tti1vna' reached timeout
2018-05-23 18:26:59,100[aceSharedPool-3]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_KJFCPWTW' with id 'workspace56avnflf4tti1vna' start failed
2018-05-23 18:27:14,351[io-8080-exec-29]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_UENKIMNB' with id 'workspace0s3skjadb1w5f5yp' created by user 'admin'
2018-05-23 18:27:20,386[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`b07e3a58-ed50-4a6e-be17-fcf49ff8b242` env=`default` workspace=`workspacegtvuzok8iyrco6kw` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-05-23T18:27:20.385591241Z`
2018-05-23 18:27:23,652[io-8080-exec-37]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_UENKIMNB' with id 'workspace0s3skjadb1w5f5yp' by user 'admin'
2018-05-23 18:30:58,600[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacehbrx3w6opa7xzbde. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacehbrx3w6opa7xzbde' reached timeout
2018-05-23 18:30:58,815[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_XLTDMOMY' with id 'workspacehbrx3w6opa7xzbde' start failed
2018-05-23 18:31:10,475[io-8080-exec-37]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 459]    - Workspace 'admin/h8k1_YQXCDZYR' with id 'workspace2fejko3rnj9myej7' created by user 'admin'
2018-05-23 18:31:19,377[io-8080-exec-24]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 314]   - Starting workspace 'admin/h8k1_YQXCDZYR' with id 'workspace2fejko3rnj9myej7' by user 'admin'
2018-05-23 18:34:58,616[aceSharedPool-2]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacegtvuzok8iyrco6kw. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacegtvuzok8iyrco6kw' reached timeout
2018-05-23 18:34:58,666[aceSharedPool-1]  [WARN ] [.i.k.KubernetesInternalRuntime 179]  - Failed to start Kubernetes runtime of workspace workspacej9cvcshgl43vwrsq. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspacej9cvcshgl43vwrsq' reached timeout
2018-05-23 18:34:58,836[aceSharedPool-2]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_SWNKKFII' with id 'workspacegtvuzok8iyrco6kw' start failed
2018-05-23 18:34:58,899[aceSharedPool-1]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 375]   - Workspace 'admin:h8k1_HYTXPZHV' with id 'workspacej9cvcshgl43vwrsq' start failed

```

The main error here seems to be `Failed to start Kubernetes runtime of workspace <workspace_id>. Cause: Waiting for Kubernetes environment 'default' of the workspace<workspace_id> reached timeout`

Im facing this issue for at least 40% to 50% of the workspaces launched this way. 

Any help on where to look and tweak stuff would be appreciated.

### Reproduction Steps

1. Set up a multi-user che with TLS enabled on Kubernetes (I used GKE)
2. Call the workspace API to create around workspaces one after the other.

**OS and version:**    

Che 6.6.0 Nightly Build

**Comment by eivantsov** [comment]
@sleshchenko i feels like the server does not get events about  agents that have successfully started. What do you think?

**Comment by sleshchenko** [comment]
@eivantsov I think there one more possible reason why workspace start can reach timeout, need to check that pods of the workspace are started. Depending on that we'll be able to plan next step of this interesting investigation =)

**Comment by eivantsov** [comment]
@mjshashank a few questions/comments:

1. Do you use the same workspace config for all workspaces?
2. Anything interesting in k8s events? I mean have all pods been created?
3. I wonder if increasing this value helps https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L454

**Comment by mjshashank** [comment]
@eivantsov and @sleshchenko thank you for pointing me in the right direction.

I have been able to dig deeper over the past couple of days into this and been able to achieve some progress.

The pods were not being created by the infra provider even though the Che logs stated the workspaces were being created. I found that this whole issue was mainly due to this [method](https://github.com/eclipse/che/blob/a2473aee7545392eaeab4b3b0e3d8f1e464db110/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceSharedPool.java#L42) which uses the the value [che.workspace.pool.exact_size](https://github.com/eclipse/che/blob/a2473aee7545392eaeab4b3b0e3d8f1e464db110/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L58) to determine how many threads are created for asynchronous workspaces tasks (eg. starting/stopping). By default this value is set to number of cores*2.

If a lot of workspaces are created, they sort of pile up since the no. of workspaces is larger than the number of threads that can be created, and before the pods are even created and the workspace start is actually carried out, a lot of them exceed the [che.infra.kubernetes.workspace_start_timeout_min](https://github.com/eclipse/che/blob/a2473aee7545392eaeab4b3b0e3d8f1e464db110/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L352) causing the above errors and workspace start failures.


The logs on the other hand do not reflect this internal queuing up and workspaces and only say `Starting workspace 'admin/workspace-name' with id 'workspace-id' by user 'admin'` before proceeding to throw the error `Failed to start Kubernetes runtime of workspace workspace-id. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspace-id' reached timeout`.

Not sure if the logs can be made more verbose and accurate but that would probably help.
